DBZ-7461 Update doc for snapshot modes alignment (#5316)
debezium/debezium: documentation/modules/ROOT/pages/connectors/mysql.adoc

|===

> [id="oracle-connector-snapshot-mode-options"]

Reviewer: ```suggestion
[id="mysql-connector-snapshot-mode-options"]
```

|`schema_only_recovery`
|Deprecated, see `recovery`.

|`recovery`
|Set this option to restore a database schema history topic that is lost or corrupted.
After a restart, the connector runs a snapshot that rebuilds the topic from the source tables.
You can also set the property to periodically prune a database schema history topic that experiences unexpected growth. +
 +
WARNING: Do not use this mode to perform a snapshot if schema changes were committed to the database after the last connector shutdown.

|`when_needed`
|The connector performs a snapshot upon startup whenever it deems it necessary. That is, when no offsets are available, or when a previously recorded offset specifies a log position that is not available in the server.

ifdef::community[]
|`custom`
|The `custom` snapshot mode lets you inject your own implementation of the `io.debezium.spi.snapshot.Snapshotter` interface.
Set the `snapshot.mode.custom.name` configuration property to the name provided by the `name()` method of your implementation, as specified on the classpath of your Kafka Connect cluster, or included in the connector JAR file, if using the `EmbeddedEngine`.

> For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].

Author: @jpechane The Snapshotter interface was marked for community only for Postgres connector, should we maintain in this way also now that is available for the other connectors?

> For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].

Reviewer 2: @mfvitale Let's keep it community only now. We'll decide if it is ready for productization base od 2.6 community feedback.

|===

[id="mysql-connector-snapshot-mode-options"]
.Settings for `snapshot.mode` connector configuration property
[cols="30%a,70%a",options="header"]
|===
|Setting |Description

|`always`

> |The connector performs a database snapshot after every restart.

Reviewer 3: ```suggestion
|The connector performs a snapshot every time that it starts.
The snapshot includes the structure and data of the captured tables.
Specify this value to populate topics with a complete representation of the data from the captured tables every time that the connector starts.
```

`minimal`:: the connector holds the global read lock for only the initial portion of the snapshot during which the connector reads the database schemas and other metadata. The remaining work in a snapshot involves selecting all rows from each table. The connector can do this in a consistent fashion by using a REPEATABLE READ transaction. This is the case even when the global read lock is no longer held and other MySQL clients are updating the database. +

`minimal_percona`:: the connector holds link:https://www.percona.com/doc/percona-server/8.0/management/backup_locks.html[the global backup lock] for only the initial portion of the snapshot during which the connector reads the database schemas and other metadata. The remaining work in a snapshot involves selecting all rows from each table. The connector can do this in a consistent fashion by using a REPEATABLE READ transaction. This is the case even when the global backup lock is no longer held and other MySQL clients are updating the database. This mode does not flush tables to disk, is not blocked by long-running reads, and is available only in Percona Server. +

> `extended`:: blocks all writes for the duration of the snapshot. Use this setting if there are clients that are submitting operations that MySQL excludes from REPEATABLE READ semantics. +

Author: I think it's better now ;-)

ifdef::community[]
`custom`:: The connector performs a snapshot according to the implementation specified by the xref:mysql-property-snapshot-locking-mode-custom-name[`snapshot.locking.mode.custom.name`] property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotLock` interface.
endif::community[]

ifdef::community[]
|[[mysql-property-snapshot-locking-mode-custom-name]]<<mysql-property-snapshot-locking-mode-custom-name, `+snapshot.locking.mode.custom.name+`>>
|No default
| When `snapshot.locking.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.
For more information, see xref:connector-custom-snapshot[custom snapshotter SPI].
endif::community[]

|[[mysql-property-snapshot-query-mode]]<<mysql-property-snapshot-query-mode, `+snapshot.query.mode+`>>
|`select_all`
|Specifies how the connector queries data while performing a snapshot. +
Set one of the following options:

> `select_all`:: The connector performs a `select all` query.

Reviewer: I'm concerned about the phrasing of "select all".  I'm worried that users may take that to imply "SELECT *" and that isn't the case when a user configures a column include/exclude configuration.  Maybe we could augment the description of this option like:

```
`select_all`:: The connector performs a `select all` query by default, optionally adjusting the columns selected based on the column include and exclude list configurations.
```

Author: Totally agreed.

|Positive integer value that specifies the number of milliseconds the connector should wait for new change events to appear before it starts processing a batch of events. Defaults to 500 milliseconds, or 0.5 second.

|[[mysql-property-snapshot-mode]]<<mysql-property-snapshot-mode, `+snapshot.mode+`>>

> |_initial_

Author: I found the same inconsistency and I don't have strong opinion on what would be better. Feel free to choose what you think is the best.

> |_initial_

Reviewer 3: For expediency, let's keep your version here for now. If I find that we have a standard, or if we decide we should implement one, we can do a full-scale sweep through all of the docs to apply consistent formatting.

|Specifies the criteria for running a snapshot when the connector starts. Possible settings are: +

> `always`:: The snapshot includes the structure and data of the captured tables.

Reviewer 3: ```suggestion
`always`:: The connector performs a snapshot every time that it starts.
The snapshot includes the structure and data of the captured tables.
```

Specify this value to populate topics with a complete representation of the data from the captured tables every time that the connector starts.

> `initial`:: The connector runs a snapshot only when no offsets have been recorded for the logical server name.

Author: I would add that  it runs also when a not completed snapshot is found.
